sql: update privilege and user account management info (#1089)
pingcap/docs: sql/privilege.md

# Privilege Management

> TiDB's privilege management system is implemented according to the privilege management system in MySQL. It supports most of the syntaxes and privilege types in MySQL.

Reviewer: I think it's important to nuance what the support looks like. TiDB is mostly compatible with MySQL 5.7, but supports roles as well (MySQL 8.0).  It does not support [dynamic privileges](https://dev.mysql.com/doc/refman/8.0/en/privileges-provided.html#privileges-provided-dynamic) (8.0) to my knowledge.

```suggestion
TiDB supports MySQL 5.7's privilege management system, including the syntax and privilege types. Starting with TiDB 3.0, support for SQL Roles is also available.
```

PTAL @tiancaiamao

Reviewer 2: Yes, you're right

| Grant          | `GrantPriv`      | Grants another user a privilege         |
| Create         | `CreatePriv`     | Creates a schema or table        |
| Select         | `SelectPriv`     | Reads the table data               |
| Insert         | `InsertPriv`     | Inserts data to a table             |
| Update         | `UpdatePriv`     | Updates the table data             |
| Delete         | `DeletePriv`     | Deleted the table data             |
| Trigger        | `TriggerPriv`    | /                 |
| Process        | `ProcessPriv`    | Displays the running task       |
| Execute        | `ExecutePriv`    | Executes the `EXECUTE` statement       |
| Drop Role      | `DropRolePriv`   | Executes `DROP ROLE`           |
| Show View      | `ShowViewPriv`   | Executes `SHOW CREATE VIEW`    |
| References     | `ReferencesPriv` | /                |
| Create View    | `CreateViewPriv` | Creates a View                 |
| Create User    | `CreateUserPriv` | Creates a user                |
| Create Role    | `CreateRolePriv` | Executes `CREATE ROLE`         |
| Show Databases | `ShowDBPriv`     | Shows the table status in the database |

### ALTER

> - For all the `ALTER` statement, users must have the `ALTER` privilege for the corresponding table.

Reviewer 3: ```suggestion
- For all `ALTER` statements, users must have the `ALTER` privilege for the corresponding table.
```

- For statements except `ALTER...DROP` and `ALTER...RENAME TO`, users must have the `INSERT` and `CREATE` privileges for the corresponding table.
- For the `ALTER...DROP` statement, users must have the `DROP` privilege for the corresponding table.
- For the `ALTER...RENAME TO` statement, users must have the `DROP` privilege for the table before renaming, and the `CREATE` and `INSERT` privileges for the table after renaming.

> **Note:**
>
> In MySQL 5.7 documentation, users need `INSERT` and `CREATE` privileges to perform the `ALTER` operation on a table. But in reality for MySQL 5.7.25, only the `ALTER` privilege is required in this case. Currently, the `ALTER` privilege in TiDB is consistent with the actual behavior in MySQL.

### CREATE DATABASE

Requires the `CREATE` privilege for the database.

### CREATE INDEX

Requires the `INDEX` privilege for the table.

### CREATE TABLE

Requires the `CREATE` privilege for the table.

To execute the `CREATE TABLE...LIKE...` statement, the `SELECT` privilege for the table is required.

### CREATE VIEW

Requires the `CREATE VIEW` privilege.

> **Note:**
>

> > If the current user is not the user that creates the View, he needs both the `CREATE VIEW` and `SUPER` privileges.

Reviewer 3: ```suggestion
> If the current user is not the user that creates the View, both the `CREATE VIEW` and `SUPER` privileges are required.
```

### DROP DATABASE

Requires the `DROP` privilege for the table.

### DROP INDEX

Requires the `INDEX` privilege for the table.

### DROP TABLES

Requires the `DROP` privilege for the table.

### TRUNCATE TABLE

Requires the `DROP` privilege for the table.

### RENAME TABLE

Requires the `ALTER` and `DROP` privileges for the table before renaming and the `CREATE` and `INSERT` privileges for the table after renaming.

### ANALYZE TABLE

Requires the `INSERT` and `SELECT` privileges for the table.

### SHOW

> `SHOW CREATE TABLE` requires any of the privileges.

Reviewer 3: Does this mean all privileges or any one of the privileges?

Author: @lilin90 Any one of the privileges.

> `SHOW CREATE TABLE` requires any of the privileges.

Reviewer: ```suggestion
`SHOW CREATE TABLE` requires any single privilege to the table.
```

`SHOW CREATE VIEW` requires the `SHOW VIEW` privilege.

### CREATE ROLE/USER

`CREATE ROLE` requires the `CREATE ROLE` privilege.

`CREATE USER` requires the `CREATE USER` privilege.

### DROP ROLE/USER

> `DROP ROLE` requires the `DROPROLE` privilege.

Reviewer 2: s/DROPROLE/DROP ROLE

> `DROP USER` requires the `CREATEUSER` privilege.

Reviewer 2: s/CREATEUSER/CREATE USER

### ALTER USER

> Requires the `CREATEUSER` privilege.

Reviewer 2: s/CREATEUSER/CREATE USER

The use of `%` in `tables_priv` and `columns_priv` is similar, but column value in `Db`, `Table_name` and `Column_name` cannot contain `%`. The sorting is also similar when loaded.

### Time of effect

> When TiDB starts, some privilege-check tables are loaded into memory, and then the cached data is used to verify the privileges. The system will periodically synchronize the `grant` table from database to cache. Time of effect is determined by the synchronization cycle. Currently, the value is 5 minutes.

Reviewer: This statement is now out of date. Using the privilege management statement interface will [synchronously run flush privileges](https://github.com/pingcap/tidb/pull/8886) (for mysql compatibility).

```suggestion
When TiDB starts, some privilege-check tables are loaded into memory, and then the cached data is used to verify the privileges. Executing privilege management statements such as `GRANT`, `REVOKE`, `CREATE USER`, `DROP USER` will take effect immediately.

Manually editing tables such as `mysql.user` with statements such as `INSERT`, `DELETE`, `UPDATE` will not take effect immediately. This behavior is compatible with MySQL, and privilege cache can be updated with the following statement:
```

PTAL @tiancaiamao
